core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -532,6 +533,23 @@ private void innerIndex(Index index) throws IOException {
         }
     }
 
+    /**
+     * Asserts that the doc in the index operation really doesn't exist
+     */
+    private boolean assertDocDoesNotExist(Index index) throws IOException {
+        final VersionValue versionValue = versionMap.getUnderLock(index.uid());
+        if (versionValue != null && versionValue.delete() == false) {
+            throw new AssertionError("doc [" + index.type() + "][" + index.id() + "] exists in version map (version " + versionValue + ")");
+        }
+        try (final Searcher searcher = acquireSearcher("assert doc doesn't exist")) {
+            final long docsWithId = searcher.reader().totalTermFreq(index.uid());
```

Comment:
this is tricky TTF will not respect deletes, I think you need to fetch the doc though. It's not much overhead compared to this.